### PR TITLE
fixes #3697, #3701 - more comprehensive field validations

### PR DESCRIPTION
**Pull request by stbenjam**
- Hostnames should be validated per RFC1123 
- Hostname shortnames should not have a period
- Upcase hostnames automatically downcased instead of outright failure
- IPv4 addresses validated that each octet is < 255

### CODE REVIEW

File: app/models/nic/base.rb [hunk 1]
```diff
@@ -10,9 +10,15 @@ class Base < ActiveRecord::Base
...
-    before_validation :normalize_mac
+    before_validation :normalize_mac,
+                      :normalize_name
 
-    validates :mac, :uniqueness => true, :presence => true, :format => {:with => Net::Validations::MAC_REGEXP}
+    validates :mac,  :uniqueness => true, 
+                     :presence => true,
+                     :format => {:with => Net::Validations::MAC_REGEXP}
+
+    validates :name, :allow_blank => true,
```
Comment (domcleal) [inline app/models/nic/base.rb]: This should be in managed.rb as we support "unmanaged" NICs (with no name) and managed with a DNS entry etc.

File: app/models/nic/base.rb [hunk 2]
```diff
@@ -54,5 +60,9 @@ def uniq_with_hosts
     def normalize_mac
       self.mac = Net::Validations.normalize_mac(mac)
     end
+
+    def normalize_name
+      self.name = Net::Validations.normalize_hostname(name) if self.name
```
Comment (domcleal) [inline app/models/nic/base.rb]: maybe use `self.name.present?` which won't then run this on empty strings either (ditto in the host above)

File: lib/net/validations.rb [hunk 3]
```diff
@@ -1,8 +1,10 @@
 module Net
   module Validations
 
-    IP_REGEXP  = /^(\d{1,3}\.){3}\d{1,3}$/
+    IP_REGEXP  = /^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\.|$)){4}$/
     MAC_REGEXP = /^([a-f0-9]{1,2}:){5}[a-f0-9]{1,2}$/i
+    HOST_REGEXP = /^(([a-z0-9]|[a-z0-9][a-z0-9\-]*[a-z0-9])\.)*([a-z0-9]|[a-z0-9][a-z0-9\-]*[a-z0-9])$/
```
Comment (domcleal) [inline lib/net/validations.rb]: Nitpick, these should use \A instead of ^ and \Z instead of $ (difference is end of string instead of end of line)

Comment (stbenjam) [inline lib/net/validations.rb]: Ah, ok, didn't realize the difference. Should be lowercase 'z',  though? 

http://stackoverflow.com/questions/577653/difference-between-a-z-and-in-ruby-regular-expressions

Comment (domcleal) [inline lib/net/validations.rb]: You're right, it should... that's an error in a6b0eeb0 too then :(

File: lib/net/validations.rb [hunk 4]
```diff
@@ -18,6 +20,12 @@ def validate_mac mac
       mac
     end
 
+    # validates the hostname
+    def validate_hostname hostname
+      raise Error, "Invalid Hostname #{hostname}" unless (hostname =~ HOST_REGEXP)
+      mac
```
Comment (GregSutcliffe) [inline lib/net/validations.rb]: Why are we returning a mac here? typo?

File: test/unit/nic_test.rb [hunk 5]
```diff
@@ -19,7 +19,7 @@ def teardown
 
   test "type casting should return the correct class" do
     i = ''
-    i = Nic::Base.create! :ip => "127.2.3.8", :mac => "babbccddeeff", :host => hosts(:one), :name => hosts(:one).name + "!", :type => "Nic::Interface"
+    i = Nic::Base.create! :ip => "127.2.3.8", :mac => "babbccddeeff", :host => hosts(:one), :name => hosts(:one).name, :type => "Nic::Interface"
```
Comment (GregSutcliffe) [inline test/unit/nic_test.rb:22]: why drop the "!"?

File: test/unit/subnet_test.rb [hunk 6]
```diff
@@ -225,6 +225,9 @@ def setup_user operation
     # missing dot
     s.network = "100101.102.103."
     refute s.valid?
+    # greater than 255
+    s.network = "300.300.300.0"
+    refute s.valid?
```
Comment (GregSutcliffe) [inline test/unit/subnet_test.rb:230]: Good idea, but I don't see any changes to the subnet model to use the new validations?

Comment (GregSutcliffe) [inline test/unit/subnet_test.rb:230]: Scratch that, I see it's in the model already :P

File: test/unit/nic_test.rb (hunk 5 above)
Comment (stbenjam) [inline test/unit/nic_test.rb:22]: Why was it there to begin with?  

The original code added a NIC with a hostname "my5name.mydomain.net!" -- which of course doesn't validate now that we're checking.

File: lib/net/validations.rb (hunk 4 above)
Comment (stbenjam) [inline lib/net/validations.rb]: Yup, copy-paste from the other method.  This is only used in ./test/lib/net/validations_test.rb, which I actually need to update to include a test for hostname validation.

File: test/unit/nic_test.rb (hunk 5 above)
Comment (GregSutcliffe) [inline test/unit/nic_test.rb:22]: Fair enough, I'm always wary of test changes that seem unrelated at first glance :)

**Comment by GregSutcliffe** [comment]
Looks ok to me, once that typo is fixed up :)

File: app/models/host/managed.rb [hunk 7]
```diff
@@ -154,6 +155,11 @@ def shortname
     domain.nil? ? name : name.chomp("." + domain.name)
```
Comment (stbenjam) [inline app/models/host/managed.rb]: This breaks: 
  https://github.com/theforeman/foreman/blob/develop/test/lib/has_many_common_test.rb#L59
1. Given a host, my5name.mydomain.net
2. Set domain directly to "yourdomain.net" (as is done in has_many_common_test)
3. shortname_has_no_periods gets called on host.save! 
4. But "name" is STILL "my5name.mydomain.net" but domain is "yourdomain.net"
5. So the "shortname" is my5name.mydomain.net"
6. Validation fails -- there's periods in the "shortname"

Any ideas here how to get name updated to reflect the new real domain?

File: app/models/host/managed.rb [hunk 8]
```diff
@@ -154,6 +155,19 @@ def shortname
...
+    errors.add(:name, "Do not include periods except in the domain name") if shortname.include? "."
+  end
+
+  def domain_name=(new_domain)
+    return if new_domain.nil?
+    # we're about to change the domain name. we need to remove
+    # the old one by setting :name to shortname
+    update_attribute(:name, shortname)
+    update_attribute(:domain, Domain.find_by_name(new_domain))
+  end
```
Comment (stbenjam) [inline app/models/host/managed.rb]: So this fixes the problem I mentioned earlier with the test (changes the domain by string).  

I don't know if this is too underhanded or bad practice to override domain_name here.  This is approaching the limit of my knowledge.

File: app/models/host/managed.rb (hunk 7 above)
Comment (domcleal) [inline app/models/host/managed.rb]: Sorry for the delay!

I think the way to solve this might be to add a before_validation method which is able to update the name when the domain changes, rather than overriding the setters (there are more than just domain_name=).

Rails provides old_\* methods like old_domain, so you can compare old_domain to domain in the before_validation, then strip off the old domain.

File: app/models/host/managed.rb [hunk 9]
```diff
@@ -154,6 +155,19 @@ def shortname
     domain.nil? ? name : name.chomp("." + domain.name)
   end
 
+  def shortname_has_no_periods
+    return if name.nil?
+    errors.add(:name, "Do not include periods except in the domain name") if shortname.include? "."
```
Comment (domcleal) [inline app/models/host/managed.rb]: Although this looks ok in the UI, it will be odd in the API etc where the full validation messages are shown as it'll do:

"name Do not include..."

Best to treat it as a sentence beginning with the field name "name", e.g. `errors.add(:name, "must not include periods..")`

Also surround the message with `_( )` so it's translated.

File: app/models/host/managed.rb [hunk 10]
```diff
@@ -154,6 +155,19 @@ def shortname
     domain.nil? ? name : name.chomp("." + domain.name)
   end
 
+  def shortname_has_no_periods
```
Comment (domcleal) [inline app/models/host/managed.rb]: I'm a bit worried about unmanaged hosts that we pick up through reports, since they might not have a domain set, so shortname might not strip off the domain - causing this validation to fail.  @GregSutcliffe?

Comment (stbenjam) [inline app/models/host/managed.rb]: Would unmanaged hosts hit this in managed.rb?

File: app/models/host/managed.rb (hunk 7 above)
Comment (stbenjam) [inline app/models/host/managed.rb]: Great, that sounds better.  I suspected there was a more rails idiomatic way to do this.

File: app/models/host/managed.rb (hunk 10 above)
Comment (domcleal) [inline app/models/host/managed.rb]: I think they're still Host::Managed instances, but with managed=false :)

Comment (GregSutcliffe) [inline app/models/host/managed.rb]: I'm not sure either. Can we add a test for that? then we'll know for sure :)

**Comment by stbenjam** [comment]
I think I've got it all sorted.  We'll see if Jenkins explodes.

https://github.com/theforeman/foreman/pull/1041/files#diff-2d53894341fd1cf84a823e3bd0855495L752 was  broken -- the name wasn't actually being updated.  I've fixed it, and added a test.  That fixed my issue.

I also only do the shortname validation on managed hosts.  Unmanaged hosts with domains we don't know about could conceivably report, right?

**Comment by domcleal** [comment]
@stbenjam indeed they could

**Comment by stbenjam** [comment]
@domcleal Ok -- so I added tests to cover that.  

It also has the side benefit of testing successful importing from facts and reports, which I don't see tested anywhere else (only the negative case was tested).

File: app/models/host/managed.rb [hunk 11]
```diff
@@ -746,13 +745,17 @@ def normalize_hostname
       # try to assign the domain automatically based on our existing domains from the host FQDN
       self.domain = Domain.all.select{|d| name.match(d.name)}.first rescue nil
     else
-      # if our host is in short name, append the domain name
+      # if we've just updated the domain name, strip off the old one
       if !new_record? and changed_attributes['domain_id'].present?
         old_domain = Domain.find(changed_attributes["domain_id"])
-        self.name.gsub(old_domain.to_s,"")
+        update_attribute(:name, self.name.chomp("." + old_domain.to_s))
```
Comment (domcleal) [inline app/models/host/managed.rb]: Just setting `self.name =` here might better, update_attribute causes an immediate DB update.

File: app/models/host/managed.rb [hunk 12]
```diff
@@ -746,13 +745,17 @@ def normalize_hostname
...
       end
+      # if our host is in short name, append the domain name
       self.name += ".#{domain}" unless name =~ /\./i
     end
+
+    # A managed host we should know the domain for; and the shortname shouldn't include a period
+    errors.add(:name, "must not include periods") if managed? and shortname.include? "."
```
Comment (domcleal) [inline app/models/host/managed.rb]: use `_("must not..")` for i18n

File: test/unit/host_test.rb [hunk 13]
```diff
@@ -131,6 +153,19 @@ class HostTest < ActiveSupport::TestCase
...
+    Setting[:create_new_host_when_facts_are_uploaded] = true
+    assert_equal true, Setting[:create_new_host_when_facts_are_uploaded]
+    old_count = Host.count
+    raw = parse_json_fixture('/facts_with_certname.json')
+    assert Host.importHostAndFacts(raw['name'], raw['facts'], raw['certname'])
+    host = Host.find_by_name('sinn1636.lan')
+    Setting[:create_new_host_when_facts_are_uploaded] =
+        Setting.find_by_name("create_new_host_when_facts_are_uploaded").default
+    assert_equal host.name, 'sinn1636.lan'
+    assert_equal old_count + 1, Host.count
```
Comment (domcleal) [inline test/unit/host_test.rb]:
```
assert_difference "Host.count", 1 do
  ...
end
```

will do the same (1's the default too, you can leave it off)

File: test/unit/usergroup_test.rb [hunk 14]
```diff
@@ -18,19 +18,19 @@ class UsergroupTest < ActiveSupport::TestCase
   end
 
   test "name is unique across user as well as usergroup" do
-    user = User.create :auth_source => auth_sources(:one), :login => "user", :mail  => "user@someware.com"
+    user = User.create :auth_source => auth_sources(:one), :login => "user", :mail  => "user@mydomain.net"
```
Comment (domcleal) [inline test/unit/usergroup_test.rb]: Is this change necessary?

File: lib/net/validations.rb [hunk 15]
```diff
@@ -45,5 +53,9 @@ def self.normalize_mac mac
           m.split("-").map { |nibble| "%02x" % ("0x" + nibble) }.join(":")
       end
     end
+
+   def self.normalize_hostname hostname
+     hostname.downcase if hostname.present?
```
Comment (domcleal) [inline lib/net/validations.rb]: It's a minor thing, but this method will return nil if hostname is "".  It might be better to write:

```
hostname.downcase! if hostname.present?
hostname
```

File: test/unit/usergroup_test.rb (hunk 14 above)
Comment (stbenjam) [inline test/unit/usergroup_test.rb]: Not anymore.

**Comment by domcleal** [comment]
Thanks @stbenjam!  Merged as 4bae5ced for Foreman 1.4.0.

Just to note, with @witlessbird's help, we changed the tests to call `@host.valid?` and check the error messages so we're specifically checking for which validations failed (and the error message) instead of just checking the host wouldn't save, just in case another validation failure would prevent it.

**Comment by stbenjam** [comment]
Oh right, :+1: thanks. I didn't think testing host.save all the way through.
